src/ray/gcs/gcs_client/global_state_accessor.h

```diff
@@ -231,10 +231,10 @@ class GlobalStateAccessor {
 
   /// Whether this client is connected to gcs server.
   bool is_connected_ GUARDED_BY(mutex_) = false;
-  std::unique_ptr<GcsClient> gcs_client_ GUARDED_BY(mutex_);
 
   std::unique_ptr<std::thread> thread_io_service_;
   std::unique_ptr<instrumented_io_context> io_service_;
+  std::unique_ptr<GcsClient> gcs_client_ GUARDED_BY(mutex_);
```

Comment:
This ordering is a bug. gcs_client must be declared earlier than io_service_, otherwise it can have destruction bug. It is because gcs_client uses some deps that rely on io service. 

I think we should probably promote a better way to define dependencies between components. The way dependencies are specified for gcs_client & periodical runner & gcs server updater is really dangerous..

Reply:
Do you mean `gcs_client` must be destroyed before `io_service_`?

Reply:
yes!